encord/http/bundle.py

```diff
@@ -0,0 +1,106 @@
+import logging
+from typing import Callable, TypeVar, Optional, List, Dict, Tuple, Generic, Generator
+from dataclasses import asdict, dataclass, field
+from functools import reduce
+
+log = logging.getLogger(__name__)
+
+T = TypeVar("T")
+R = TypeVar("R")
+
+
+@dataclass
+class BundleResultHandler(Generic[T]):
+    predicate: str
+    continuation: Callable[[T], None]
+
+
+@dataclass
+class BundledOperation(Generic[T, R]):
+    operation: Callable[..., List[R]]
+    reducer: Callable[[T, T], T]
+    mapper: Optional[Callable[[R], str]]
+    limit: int
+    payloads: list = field(default_factory=lambda: [])
+    continuations: Dict[str, Callable[[T], None]] = field(default_factory=lambda: dict())
+
+    def append(self, payload: T, continuation: Optional[BundleResultHandler]):
+        self.payloads.append(payload)
+        if continuation is not None:
+            self.continuations[continuation.predicate] = continuation.continuation
+
+    def get_bundled_payload(self) -> Generator[T, None, None]:
+        for i in range(0, len(self.payloads), self.limit):
+            yield reduce(self.reducer, self.payloads[i : i + self.limit])
+
+
+class Bundle:
+    """
+    This class allows to perform operations in bundles to improve performance by reducing number of network calls.
+
+    To execute batch you can either call  :meth:`.execute()` directly, or use a context manager.
+
+        .. code::
+
+                # Code example of performing batched label initialisation
+                project = ... # assuming you already have instantiated this Project object
+                label_rows = project.list_label_rows_v2()
+                with project.create_bundle() as bundle:
+                    for label_row in label_rows:
+                        label_row.initialise_labels(bundle=bundle)
+
+                # At this point all labels will be initialised
+    """
+
+    def __init__(self) -> None:
+        self._operations: Dict[Callable, BundledOperation] = dict()
+
+    def _register_operation(
+        self,
+        operation: Callable[..., List[R]],
+        reducer: Callable[[T, T], T],
+        mapper: Optional[Callable[[R], str]],
+        limit: int,
+    ) -> BundledOperation[T, R]:
+        if operation not in self._operations:
+            self._operations[operation] = BundledOperation(operation, reducer, mapper, limit)
+
+        return self._operations[operation]
+
+    def add(
+        self,
+        operation: Callable[..., List[R]],
+        reducer: Callable[[T, T], T],
+        mapper: Optional[Tuple[Callable[[R], str], BundleResultHandler[R]]],
```

Comment:
small thing: maybe a dataclass or a namedtuple would be a bit easier to read.

Reply:
Yeah I'm very confused here given that we unpack it below as `mapper[0]` being the `mapper_predicate` and `mapper[1]` being the `mapper_continuation` but above the `BundleResultHandler` which to me is `mapper[1]` here has a `predicate` and `continuation` field as well.